aten/src/ATen/native/cuda/int4mm.cu

```diff
@@ -0,0 +1,1020 @@
+
+  TORCH_CHECK(
+      KTilesPerWarp >= BLayout::kInnerKTiles && isPowerOf2(KTilesPerWarp));
+
+  // After intra-block reduction across the k dimension, we are left with this
+  // many tiles
+  //  int32_t postKernelKTiles = kTiles / (Warps * KTilesPerWarp);
+  int32_t postKernelKTiles = 1; // we loop
+
+  auto grid = dim3(postKernelKTiles, nTiles, mTiles);
+  auto block = dim3(kWarpSize, Warps);
+
+  auto func =
+      tinygemm_m16n8k16_chunk_kernel<ALayout, BLayout, CLayout, Warps, KTilesPerWarp>;
+
+  func<<<grid, block, 0, stream>>>(
+      A.data_ptr(),
+      B.data_ptr(),
+      qScaleAndZeros ? qScaleAndZeros->data_ptr() : nullptr,
+      C_final.data_ptr(),
+      m,
+      n,
+      k,
+      mTiles,
+      nTiles,
+      kTiles);
+
+  cudaFuncAttributes funcAttr;
+  C10_CUDA_CHECK(cudaFuncGetAttributes(
+      &funcAttr,
+      func));
+}
+
+// FIXME: parallelize better, smem staging etc?
+template <int InnerKTiles>
+__global__ void matrix_to_m16n8k16_Bint4_layout(
+    // size [n][k]
+    const at::PackedTensorAccessor32<int32_t, 2, at::RestrictPtrTraits> in,
+    // size [ceil(n / 8)][ceil(k / (InnerKTiles * 16))][32][InnerKTiles / 2]
+    at::PackedTensorAccessor32<int32_t, 4, at::RestrictPtrTraits> out) {
+  // int4 values are packed into int32 values, which require at least 8. Given
+  // m16n8k16 B layout requires 4 scalar values/lane, the minimum number of
+  // innermost k-tiles that we can use is 2.
+  static_assert(InnerKTiles >= 2 && isPowerOf2(InnerKTiles), "");
+
+  constexpr int32_t kMTileSize = 16;
+  constexpr int32_t kNTileSize = 8;
+  constexpr int32_t kKTileSize = 16;
+
+  // gridDim.x corresponds to the number of k-tiles divided by InnerKTiles
+  auto kOuterTile = blockIdx.x;
+  auto nTile = blockIdx.y;
+  auto t = threadIdx.x;
+
+  // Two k-tiles are packed into an int32 at a time
+#pragma unroll
+  for (int innerKTile = 0; innerKTile < InnerKTiles; innerKTile += 2) {
+    // n dimension that this lane loads from
+    auto n0 = nTile * kNTileSize + (t / 4);
+
+    bool n0Valid = n0 < in.size(0);
+
+    int32_t ks[8];
+
+    auto kBase0 = (kOuterTile * InnerKTiles + innerKTile) * kKTileSize;
+    ks[0] = kBase0 + (t % 4) * 2;
+    ks[1] = ks[0] + 1;
+    ks[2] = ks[0] + 8;
+    ks[3] = ks[0] + 8 + 1;
+
+    auto kBase1 = kBase0 + kKTileSize;
+    ks[4] = kBase1 + (t % 4) * 2;
+    ks[5] = ks[4] + 1;
+    ks[6] = ks[4] + 8;
+    ks[7] = ks[4] + 8 + 1;
+
+    auto pIn = &in[n0][0];
+
+    uint32_t v[8];
+#pragma unroll
+    for (int i = 0; i < 8; ++i) {
+      v[i] = (n0Valid && ks[i] < in.size(1)) ? pIn[ks[i]] : uint32_t(0);
+    }
+
+    int32_t pack = (v[7] << 28) | (v[5] << 24) | (v[3] << 20) | (v[1] << 16) |
+        (v[6] << 12) | (v[4] << 8) | (v[2] << 4) | v[0];
+
+    // inner k-tiles pack two at a time
+    out[nTile][kOuterTile][t][innerKTile / 2] = pack;
+  }
+}
+
+#endif
+
+
+torch::Tensor _weight_int4pack_mm_cuda(
+    const torch::Tensor& A,
+    const torch::Tensor& B,
+    long qGroupSize,
```

Comment:
`sizeof(long) == 4` on Win64, but `8` on all other OSes. `sizeof(int64_t)` is unambiguously `8`
```suggestion
    int64_t qGroupSize,
```